File: Gfg/gold_mine.py
# Desccription Gold Mine Algorithm
# Given a gold mine of n*m dimensions.
# Each field in this mine contains a positive integer which is the amount of gold in tons.
# Initially the miner is at first column but can be at any row.
# He can move only (right->,right up /,right down\) that is from a given cell,
# the miner can move to the cell diagonally up towards the right or right or diagonally down towards the right.
# The program calculates the maximum amount of gold he can collect.

def max_gold(gold,m,n):
    # Table for storing intermediate results and initializing all cells to 0.
    # The first row of goldMineTable gives the maximum gold that the miner can collect when starts that row
    goldTable = [[0 for i in range(n)]
                        for j in range(m)]
 
    for col in range(n-1, -1, -1):
        for row in range(m):
 
            # Gold collected on going to the cell on the rigth(->)
            if (col == n-1):
                right = 0
            else:
                right = goldTable[row][col+1]
 
            # Gold collected on going to the cell to right up (/)
            if (row == 0 or col == n-1):
                right_up = 0
            else:
                right_up = goldTable[row-1][col+1]
 
            # Gold collected on going to the cell to right down (\)
            if (row == m-1 or col == n-1):
                right_down = 0
            else:
                right_down = goldTable[row+1][col+1]
 
            # Max gold collected from taking either of the above 3 paths
            goldTable[row][col] = gold[row][col] + max(right, right_up, right_down)
                                                            
    # The max amount of gold collected will be the max value in first column of all rows
    res = goldTable[0][0]
    for i in range(1, m):
        res = max(res, goldTable[i][0])
 
    return res
    

# Main function
def main():
    

    m = int(input('Enter number of rows : '))
    n = int(input('Enter number of columns : '))

    print('Enter gold values')
    gold_mat = [[int(j) for j in input().split()]for i in range(n)]

    print('GOLD MINE MATRIX :')
    for i in range(len(gold_mat)):
        for j in range(len(gold_mat[i])):
            print(gold_mat[i][j],end=" ")
        print()
    
    # To calculate maximum gold 
    print('The maximum amount of gold collected : ',max_gold(gold_mat,m,n))

# To check if the file contains the main function and then calling the main function
if __name__ == '__main__':
    main()

# COMPLEXITY:
# Time Complexity :O(m*n) 
# Space Complexity :O(m*n)

# OUTPUT
            # Input  : 
                # gold_mat = 10 9 2 30
                          #  15 12 3 40
                          #  11 8 4 50
            # Output : 
                # Maximum amount of gold collected : 81

File: Algorithms/Dynamic Problems/C++/Gold_mine.cpp
/*
 Problem : We are given a gold mine of n*m dimensions.
 Each field in mine contains a positive integer which is the amount of gold in tons. Initially the miner can start from any row in the first column. From a given cell, the miner can move

 1. to the cell diagonally up towards the right 
 2. to the right
 3. to the cell diagonally down towards the right
 we need to find out maximum amount of gold which he can collect.

 Sample I/p : 4 4
              1 3 1 5 2 2 4 1 5 0 2 3 0 6 1 2
 Sample O/P : 16
 */

 #include <bits/stdc++.h>
 using namespace std;

 int maxGold(int n, int m, vector<vector<int> > gold_mine) {
         int ans = INT_MIN;
         vector<vector<int> > dp(n+2, vector<int> (m+1, 0));
         for(int j = 1;  j <= m; j++){
             for(int i = 1; i <= n; i++){
                 dp[i][j] = max(dp[i-1][j-1], max(dp[i][j-1], dp[i+1][j-1])) + gold_mine[i-1][j-1];
             }
         }

         for(int i = 0;i <= n;i++)
             ans = max(dp[i][m], ans);

         return ans;
 }

 int main() {
     int m,n;
     cin >> n >> m;
     vector<vector<int> > gold_mine(n, vector<int>(m, 0));
     for (int i = 0; i < n; i++) {
         for (int j = 0; j < m; j++) {
             cin >> gold_mine[i][j];
         }
     }

     cout<< maxGold(n, m, gold_mine) <<"\n";
     return 0;

 } 